### ISSUE TITLE
[Bug?]: Topological from multiple workspaces breaks parallelisation

### ISSUE THREAD

**Issue by samhh** [yarnpkg/berry#6106]
### Self-service

- [ ] I'd be willing to implement a fix

### Describe the bug

You can find a small repro here: https://git.sr.ht/~samhh/yarn-repro. I'll copy its README here:

---

There are three workspaces: `a`, `b`, and `c`. `b` depends upon `c`.

Repro:

```console
$ yarn run repro
```

Expected output order:

```
c
b
a
```

`a` has no dependencies, however it's sleeping the longest, so it should come last.

Actual output order:

```
c
a
b  
```

`b` doesn't start until both `a` and `c` have finished, despite the fact we passed `-p` with `--from '{a,b}'`.

This behaviour is inconsistent, as `a` and `b` would have no dependency upon one another if `b` didn't depend upon `c`. Patch to test that with:

```patch
diff --git a/workspaces/b/package.json b/workspaces/b/package.json
index d89b747..559721c 100644
--- a/workspaces/b/package.json
+++ b/workspaces/b/package.json
@@ -2,8 +2,5 @@
   "name": "b",
   "scripts": {
     "echo": "echo b"
-  },
-  "dependencies": {
-    "c": "workspace:*"
   }
 }
diff --git a/yarn.lock b/yarn.lock
index 1cc71b4..d6edac7 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -14,12 +14,10 @@ __metadata:
 "b@workspace:workspaces/b":
   version: 0.0.0-use.local
   resolution: "b@workspace:workspaces/b"
-  dependencies:
-    c: "workspace:*"
   languageName: unknown
   linkType: soft
 
-"c@workspace:*, c@workspace:workspaces/c":
+"c@workspace:workspaces/c":
   version: 0.0.0-use.local
   resolution: "c@workspace:workspaces/c"
   languageName: unknown

```

### To reproduce

Full repro with README here: https://git.sr.ht/~samhh/yarn-repro

In short, take this `package.json`:

```json
{
  "name": "yarn-repro",
  "packageManager": "yarn@4.0.2",
  "workspaces": [
    "./workspaces/*"
  ],
  "scripts": {
    "repro": "yarn workspaces foreach -Rpit --from '{a,b}' run echo"
  }
}
```

And add `:/workspaces/{a,b,c}/` with `b` depending upon `c`, and add `echo` scripts with `sleep` as necessary to exhibit the behaviour.

### Environment

```shell
System:
    OS: macOS 14.2.1
    CPU: (12) arm64 Apple M2 Pro
  Binaries:
    Node: 20.10.0 - /private/var/folders/32/3g74tn455yq7bczhtrlttjnr0000gn/T/xfs-e0fd97bd/node
    Yarn: 4.0.2 - /private/var/folders/32/3g74tn455yq7bczhtrlttjnr0000gn/T/xfs-e0fd97bd/yarn
    npm: 10.2.3 - /nix/store/m8s1dj96nbackrk4gk1bsn8hlfcggg4b-nodejs-20.10.0/bin/npm
```
